### ISSUE TITLE
[Bug]: [ v0.11,v0.18.rc1]A general accuracy bug report and solution discussion, theoretically involving all MOE models including the Minimax series and Qwen series.

### ISSUE THREAD

**Issue by florenceCH** [vllm-project/vllm-ascend#8359]
### Your current environment

<details> 0.11 and 0.18.RC1 A2 910B3 
<summary>The output of `python collect_env.py`</summary>

```text
Collecting environment information...
PyTorch version: 2.9.0+cpu
Is debug build: False

OS: Ubuntu 22.04.5 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 4.3.1
Libc version: glibc-2.35

Python version: 3.11.10 (main, Apr  8 2026, 15:39:58) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.10.0-136.12.0.86.h1903.eulerosv2r12.aarch64-aarch64-with-glibc2.35

CPU:
Architecture:                         aarch64
CPU op-mode(s):                       64-bit
Byte Order:                           Little Endian
CPU(s):                               192
On-line CPU(s) list:                  0-191
Vendor ID:                            HiSilicon
BIOS Vendor ID:                       HiSilicon
Model name:                           Kunpeng-920
BIOS Model name:                      HUAWEI Kunpeng 920 5250
Model:                                0
Thread(s) per core:                   1
Core(s) per socket:                   48
Socket(s):                            4
Stepping:                             0x1
BogoMIPS:                             200.00
Flags:                                fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma dcpop asimddp asimdfhm ssbs
L1d cache:                            12 MiB (192 instances)
L1i cache:                            12 MiB (192 instances)
L2 cache:                             96 MiB (192 instances)
L3 cache:                             192 MiB (8 instances)
NUMA node(s):                         8
NUMA node0 CPU(s):                    0-23
NUMA node1 CPU(s):                    24-47
NUMA node2 CPU(s):                    48-71
NUMA node3 CPU(s):                    72-95
NUMA node4 CPU(s):                    96-119
NUMA node5 CPU(s):                    120-143
NUMA node6 CPU(s):                    144-167
NUMA node7 CPU(s):                    168-191
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; __user pointer sanitization
Vulnerability Spectre v2:             Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pyzmq==27.1.0
[pip3] torch==2.9.0
[pip3] torch_npu==2.9.0
[pip3] torchaudio==2.9.0
[pip3] torchvision==0.24.0
[pip3] transformers==4.57.6
[pip3] triton-ascend==3.2.0
[conda] Could not collect
vLLM Version: 0.1.dev14925+g10bf4cd23.d20260408 (git sha: 10bf4cd23, date: 20260408)
vLLM Ascend Version: 0.1.dev2799+gfbce1b740 (git sha: fbce1b740)

ENV Variables:
ATB_OPSRUNNER_KERNEL_CACHE_TILING_SIZE=10240
ATB_OPSRUNNER_KERNEL_CACHE_LOCAL_COUNT=1
ATB_STREAM_SYNC_EVERY_RUNNER_ENABLE=0
ATB_OPSRUNNER_SETUP_CACHE_ENABLE=1
ATB_WORKSPACE_MEM_ALLOC_GLOBAL=0
ATB_DEVICE_TILING_BUFFER_BLOCK_NUM=32
ATB_STREAM_SYNC_EVERY_KERNEL_ENABLE=0
ATB_OPSRUNNER_KERNEL_CACHE_GLOABL_COUNT=5
ATB_HOME_PATH=/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_1
ASCEND_TOOLKIT_HOME=/usr/local/Ascend/cann-8.5.1
ATB_COMPARE_TILING_EVERY_KERNEL=0
ASCEND_OPP_PATH=/usr/local/Ascend/cann-8.5.1/opp
LD_LIBRARY_PATH=/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_0/lib:/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_0/examples:/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_0/tests/atbopstest:/usr/local/Ascend/ascend-toolkit/latest/tools/aml/lib64:/usr/local/Ascend/ascend-toolkit/latest/tools/aml/lib64/plugin:/usr/local/Ascend/ascend-toolkit/latest/lib64:/usr/local/Ascend/ascend-toolkit/latest/lib64/plugin/opskernel:/usr/local/Ascend/ascend-toolkit/latest/lib64/plugin/nnengine:/usr/local/Ascend/ascend-toolkit/latest/opp/built-in/op_impl/ai_core/tbe/op_tiling:/usr/local/Ascend/driver/lib64/common/:/usr/local/Ascend/driver/lib64/driver/:/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_0/lib:/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_0/examples:/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_0/tests/atbopstest:/usr/local/Ascend/ascend-toolkit/latest/tools/aml/lib64:/usr/local/Ascend/ascend-toolkit/latest/tools/aml/lib64/plugin:/usr/local/Ascend/ascend-toolkit/latest/lib64/plugin/nnengine:/usr/local/Ascend/ascend-toolkit/latest/opp/built-in/op_impl/ai_core/tbe/op_tiling/lib/linux/:/usr/local/Ascend/ascend-toolkit/latest/lib64:/usr/local/Ascend/ascend-toolkit/latest/lib64/plugin/opskernel:/usr/local/Ascend/driver/lib64/driver:/usr/local/Ascend/driver/lib64/common:/usr/local/lib::/usr/local/Ascend/driver/lib64/driver:/usr/local/Ascend/driver/lib64/common:/usr/local/lib::/usr/local/Ascend/ascend-toolkit/latest/aarch64-linux/devlib
ASCEND_AICPU_PATH=/usr/local/Ascend/cann-8.5.1
ATB_OPSRUNNER_KERNEL_CACHE_TYPE=3
ATB_RUNNER_POOL_SIZE=64
ATB_STREAM_SYNC_EVERY_OPERATION_ENABLE=0
ASCEND_HOME_PATH=/usr/local/Ascend/cann-8.5.1
ATB_MATMUL_SHUFFLE_K_ENABLE=1
ATB_LAUNCH_KERNEL_WITH_TILING=1
ATB_WORKSPACE_MEM_ALLOC_ALG_TYPE=1
ATB_HOST_TILING_BUFFER_BLOCK_NUM=128
ATB_SHARE_MEMORY_NAME_SUFFIX=
TORCH_DEVICE_BACKEND_AUTOLOAD=1
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1


NPU:
+------------------------------------------------------------------------------------------------+
| npu-smi 25.2.0                   Version: 25.2.0                                               |
+---------------------------+---------------+----------------------------------------------------+
| NPU   Name                | Health        | Power(W)    Temp(C)           Hugepages-Usage(page)|
| Chip                      | Bus-Id        | AICore(%)   Memory-Usage(MB)  HBM-Usage(MB)        |
+===========================+===============+====================================================+
| 0     910B3               | OK            | 98.6        47                0    / 0             |
| 0                         | 0000:C1:00.0  | 0           0    / 0          61339/ 65536         |
+===========================+===============+====================================================+
| 1     910B3               | OK            | 100.5       48                0    / 0             |
| 0                         | 0000:01:00.0  | 0           0    / 0          61330/ 65536         |
+===========================+===============+====================================================+
| 2     910B3               | OK            | 96.4        47                0    / 0             |
| 0                         | 0000:C2:00.0  | 0           0    / 0          61330/ 65536         |
+===========================+===============+====================================================+
| 3     910B3               | OK            | 98.1        47                0    / 0             |
| 0                         | 0000:02:00.0  | 0           0    / 0          61330/ 65536         |
+===========================+===============+====================================================+
| 4     910B3               | OK            | 97.8        47                0    / 0             |
| 0                         | 0000:81:00.0  | 0           0    / 0          61330/ 65536         |
+===========================+===============+====================================================+
| 5     910B3               | OK            | 106.4       50                0    / 0             |
| 0                         | 0000:41:00.0  | 0           0    / 0          61330/ 65536         |
+===========================+===============+====================================================+
| 6     910B3               | OK            | 97.3        46                0    / 0             |
| 0                         | 0000:82:00.0  | 0           0    / 0          61330/ 65536         |
+===========================+===============+====================================================+
| 7     910B3               | OK            | 97.0        49                0    / 0             |
| 0                         | 0000:42:00.0  | 0           0    / 0          61330/ 65536         |
+===========================+===============+====================================================+
+---------------------------+---------------+----------------------------------------------------+
| NPU     Chip              | Process id    | Process name             | Process memory(MB)      |
+===========================+===============+====================================================+
| 0       0                 | 58457         |                          | 58029                   |
+===========================+===============+====================================================+
| 1       0                 | 58458         |                          | 58028                   |
+===========================+===============+====================================================+
| 2       0                 | 58459         |                          | 58029                   |
+===========================+===============+====================================================+
| 3       0                 | 58460         |                          | 58028                   |
+===========================+===============+====================================================+
| 4       0                 | 58462         |                          | 58029                   |
+===========================+===============+====================================================+
| 5       0                 | 58491         |                          | 58029                   |
+===========================+===============+====================================================+
| 6       0                 | 58492         |                          | 58029                   |
+===========================+===============+====================================================+
| 7       0                 | 58493         |                          | 58029                   |
+===========================+===============+====================================================+

CANN:
package_name=Ascend-cann-toolkit
version=8.5.1
innerversion=V100R001C25SPC002B220
compatible_version=[V100R001C15],[V100R001C18],[V100R001C19],[V100R001C20],[V100R001C21],[V100R001C23]
arch=aarch64
os=linux
**path=/usr/local/Ascend/cann-8.5.1**


```

</details>

### 🐛 Describe the bug

We built coding service using vllm-ascend **v0.11 and v0.18.rc1** as the foundation. The models used were the **MINIMAX M2 series and Qwen3.5 series**. During the task process, we found that in complex coding and agent tasks, there were **occasional accuracy issues** （**about 1% probability**） such as loss of toolcall and think tags, repeated outputs, and decreased model compliance.

```text
part of badcase output examples:

\n1.  步骤：查询结果展示</parameter>\nname=\"skill\">## Skill完成目标\n\n本Skill的目标是完成以下任务\n\n 1. 严格按照步骤完成。\n\n本步骤1\n\n 1. 严格按照步骤完成。\n\n2. 严格遵守每一步骤。\n\n2. 严格遵守每一步骤。\n\n2. 严格遵守每步骤。\n\n---\n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## 步骤完成步骤完成步骤完成步骤完成步骤完成步骤\n\n---\n\n## \n\n## 步骤\n\n 1. \n\n 1. \n\n 1. \n 1. \n\n 。\n\n### 步骤\n\n## 步骤\n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## \n\n## 步骤完成步骤\n\n---\n\n## \n\n## \n\n## \n\n## \n\n\n\n---</parameter>\n<parameter name=\"skill\">## 

```

After initial localization, we observed that there are occasional, significant fluctuations (0.5~1) in the logits after each first token inference. This fluctuation is clearly abnormal, prompting us to conduct further checks on the operator behavior.

Through deeper analysis, we found that the expert selection of the fused operator moe_gate_topk exhibits significant fluctuations. The root cause lies in the abnormal output of the upstream allreduce communication operator. Therefore, we conducted a series of experiments.

We discovered:
1. After replacing the **moe_gate_topk**  with  small operations, the probability of model accuracy degradation was significantly reduced.
2. After **enabling deterministic computation（export HCCL_DETERMINISTIC=true）**, the model's accuracy was tested with over 3000 badcase test cases and no issues were reproduced (previously 5%).

As a result, we suspect that there may be an anomaly in the underlying communication of vllm-ascend, which could lead to occasional accuracy issues.

Since deterministic computation may cause noticeable performance degradation, we will avoid using this method to mitigate accuracy issues whenever possible. In the next step, we will disable the hierarchical communication feature (**HCCL_INTRA_ROCE_ENABLE=1**) for further verification and in-depth analysis of the communication behavior.

Contact me for further discussion
email：gaoxiang120@huawei.com

**Comment by florenceCH** [comment]
The deployment mode is A2 910B3, single-node system, multi-node system with mixed deployment, or PD separation, and floating-point weights are used.

**Comment by florenceCH** [comment]
**ENV config:** 
# 服务可配置变量
# ========================
export VLLM_VERSION=0.18
MASTER_IP_ADDRESS=
IP_ADDRESS=
SERVICE_PORT=8700


# 分层调度
export HCCL_INTRA_PCIE_ENABLE=1
export HCCL_INTRA_ROCE_ENABLE=0

# 卡相关
NETWORK_INTERFACE="eth0"
export ASCEND_RT_VISIBLE_DEVICES=0,1,2,3,4,5,6,7

# 主节点为0，从节点依次递增
DECODE_DATA_PARALLEL_RANK=0

PREFILL_DATA_PARALLEL_SIZE=2
PREFILL_TENSOR_PARALLEL_SIZE=8

DECODE_DATA_PARALLEL_SIZE=2
DECODE_TENSOR_PARALLEL_SIZE=8


# ========================
# 模型配置变量
# ========================
MODEL_PATH="/opt/MiniMax-M2.7/MiniMax-M2.7/"  # 模型权重路径
MODEL_NAME="MiniMax-M2.7"  # 服务模型名称
TOOL_CALL_PARSER="minimax_m2" # 工具调用
REASONING_PARSER="minimax_m2" # 思考解析

LIBJEMALLOC_SO_PATH="/usr/lib/aarch64-linux-gnu/libjemalloc.so.2"
GPU_MEMORY_USE=0.95

# chunked-size大小
MAX_BATCHED_TOKENS=64
ADDITIONAL_CONFIG='{"ascend_scheduler_config":{"enabled": false, "enable_chunked_prefill":true}}'

# 开启MC2算子需要配置适当的HCCL_BUFFSIZE大小
export HCCL_BUFFSIZE=512


# ========================
# 功能参数（公共）
# ========================
export VLLM_USE_V1=1


# ========================
# 性能优化参数（公共）
# ========================
export OMP_NUM_THREADS=100
export OMP_PROC_BIND=false
export TASK_QUEUE_ENABLE=1    # 是否开启算子下发优化：0关闭；1或者未配置走Level 1优化；2走Level 2优化
export LD_PRELOAD=$LIBJEMALLOC_SO_PATH:$LD_PRELOAD
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True


# ========================
# 配置参数（公共）
# ========================
export HCCL_IF_IP="$IP_ADDRESS"
export GLOO_SOCKET_IFNAME="$NETWORK_INTERFACE"
export TP_SOCKET_IFNAME="$NETWORK_INTERFACE"
export HCCL_SOCKET_IFNAME="$NETWORK_INTERFACE"
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True

**Comment by weijinqian0** [comment]
@SparrowMu Plz see this issue.
